src/System.Net.Http/src/System/Net/Http/Unix/CurlHandler.MultiAgent.cs

```diff
@@ -0,0 +1,658 @@
+                        unsafe
+                        {
+                            Interop.libcurl.curl_waitfd extraFds = new Interop.libcurl.curl_waitfd {
+                                fd = _wakeupRequestedPipeFd,
+                                events = Interop.libcurl.CURL_WAIT_POLLIN,
+                                revents = 0
+                            };
+                            ThrowIfCURLMError(Interop.libcurl.curl_multi_wait(multiHandle, &extraFds, 1, FailsafeTimeoutMilliseconds, out numFds));
+                            if ((extraFds.revents & Interop.libcurl.CURL_WAIT_POLLIN) != 0)
+                            {
+                                // We woke up (at least in part) because a wake-up was requested.  Read the data out of the pipe 
+                                // to clear it. It's possible but unlikely that there will be tons of extra data in the pipe, 
+                                // more than we end up reading out here (it's unlikely because we only write a byte to the pipe when 
+                                // transitioning from 0 to 1 incoming request or when cancellation is requested, and that would
+                                // need to happen many times in a single iteration).  In that unlikely case, we'll simply loop 
+                                // around again as normal and end up waking up spuriously from the next curl_multi_wait.  For now, 
+                                // this is preferable to making additional syscalls to poll and read from the pipe).
+                                const int ClearBufferSize = 4096; // some sufficiently large size to clear the pipe in any normal case
+                                byte* clearBuf = stackalloc byte[ClearBufferSize];
+                                while (Interop.CheckIo((long)Interop.libc.read(_wakeupRequestedPipeFd, clearBuf, (IntPtr)ClearBufferSize)));
+                            }
+                        }
+
+                        // PERF NOTE: curl_multi_wait uses poll (assuming it's available), which is O(N) in terms of the number of fds 
+                        // being waited on. If this ends up being a scalability bottleneck, we can look into using the curl_multi_socket_* 
+                        // APIs, which would let us switch to using epoll by being notified when sockets file descriptors are added or 
+                        // removed and configuring the epoll context with EPOLL_CTL_ADD/DEL, which at the expense of a lot of additional
+                        // complexity would let us turn the O(N) operation into an O(1) operation.  The additional complexity would come
+                        // not only in the form of additional callbacks and managing the socket collection, but also in the form of timer
+                        // management, which is necessary when using the curl_multi_socket_* APIs and which we avoid by using just
+                        // curl_multi_wait/perform.
```

Comment:
I think the PR can go ahead and we will try to get data on whether curl_multi_socket_action+poll is better than curl_multi_perform+curl_multi_wait. Since only benefit of former seems to be that it works with a specific fd, we need to investigate if this is actually translating to O(n) vs 2*O(n) curl_multi_perform might have some inbuilt optimizations of its own